zebra-chain/src/equihash_solution.rs

```diff
@@ -15,24 +18,87 @@ const EQUIHASH_SOLUTION_SIZE: usize = 1344;
 ///
 /// The size of an Equihash solution in bytes is always 1344 so the
 /// length of this type is fixed.
-#[derive(Clone)]
-pub struct EquihashSolution([u8; EQUIHASH_SOLUTION_SIZE]);
+pub struct EquihashSolution(pub [u8; EQUIHASH_SOLUTION_SIZE]);
 
-impl Default for EquihashSolution {
-    fn default() -> Self {
-        EquihashSolution([0; EQUIHASH_SOLUTION_SIZE])
+impl PartialEq<EquihashSolution> for EquihashSolution {
+    fn eq(&self, other: &EquihashSolution) -> bool {
+        self.0.as_ref() == other.0.as_ref()
     }
 }
 
-// TODO: fix hex crate import conflict with tracing-subscriber
-// impl fmt::Debug for EquihashSolution {
-//     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-//         f.write_str(&self.0.encode_hex::<String>())
-//     }
-// }
+impl fmt::Debug for EquihashSolution {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("EquihashSolution")
+            .field(&hex::encode(&self.0[..]))
+            .finish()
+    }
+}
+
+// These impls all only exist because of array length restrictions.
+
+impl Copy for EquihashSolution {}
+
+impl Clone for EquihashSolution {
+    fn clone(&self) -> Self {
+        let mut bytes = [0; EQUIHASH_SOLUTION_SIZE];
+        bytes[..].copy_from_slice(&self.0[..]);
+        Self(bytes)
+    }
+}
+
+impl Eq for EquihashSolution {}
+
+impl ZcashSerialize for EquihashSolution {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), SerializationError> {
+        writer.write_all(&self.0[..])?;
```

Comment:
I think this needs to write a compactsize-encoded `1344` before writing the solution, per https://zips.z.cash/protocol/protocol.pdf#blockheader, since when the header is serialized it just calls `solution.zcash_serialize`.